### PR TITLE
New proposal for relaxing Paginate argument

### PR DESCRIPTION
**Pull request by vdanjean**
This proposal is discussed in issue #4865 (with numerous comments in closed PR #4866)

### CODE REVIEW

**Comment by kaushalmodi** [comment]
My only comment would be to include doc updates in this PR too, if there are plans to merge it.

File: hugolib/pagination.go [hunk 1]
```diff
@@ -414,6 +418,39 @@ func paginatePages(td targetPathDescriptor, seq interface{}, pagerSize int) (pag
...
+		return seq.(PagesGroup), nil
+	case []PageGroup:
+		return PagesGroup(seq.([]PageGroup)), nil
+	case []interface{}:
+		list := seq.([]interface{})
+		l := len(list)
+		if l == 0 {
+			break
+		}
+		switch list[0].(type) {
```
Comment (bep) [inline hugolib/pagination.go]: `switch v := list[0].(type)` same comment as above -- avoids casting inside the switches.

File: hugolib/pagination.go [hunk 2]
```diff
@@ -414,6 +418,39 @@ func paginatePages(td targetPathDescriptor, seq interface{}, pagerSize int) (pag
...
+		case PageGroup:
+			pagesGroup := make([]PageGroup, len(list))
```
Comment (bep) [inline hugolib/pagination.go]: ` make(PagesGroup, len(list))` is better.

File: hugolib/pageGroup.go [hunk 3]
```diff
@@ -296,3 +296,10 @@ func (p Pages) GroupByParamDate(key string, format string, order ...string) (Pag
 	}
 	return p.groupByDateField(sorter, formatter, order...)
 }
+
+// Group creates a PageGroup from a key and a Pages object
+func (p *Page) Group(key interface{}, pages interface{}) (PageGroup, error) {
```
[the remark this answers is not in the document]
Comment (vdanjean) [reply hugolib/pageGroup.go]: Done
But I do not really understand when using interface{} or not (when the argument can have various real types?)

File: hugolib/pagination.go (hunk 1 above)
Comment (vdanjean) [reply hugolib/pagination.go]: Not done as the value of the first element in the array is not important. Perhaps the check is not the good one.
There is a loop just after, iterating over all elements, to cast them one by one. The initial switch is here to handle type errors but perhaps it should be done for all elements.

If there is a better way to do it in go, tell me or fix it directly.

File: hugolib/pagination.go (hunk 2 above)
Comment (vdanjean) [reply hugolib/pagination.go]: Done

**Pull request merged**